src/gateway/adhoc.cpp

```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
-    remote_map_t::const_iterator lb, ub;
-
-    auto ptr = m_remotes.synchronize();
-
-    if(!ptr->count(name)) {
-        throw std::system_error(error::service_not_available);
-    }
-
-    std::tie(lb, ub) = ptr->equal_range(name);
-
-    std::uniform_int_distribution<int> distribution(0, std::distance(lb, ub) - 1);
-    std::advance(lb, distribution(m_random_generator));
-
-    COCAINE_LOG_DEBUG(m_log, "providing service using remote actor", attribute_list({
-        {"uuid", lb->second.uuid}
-    }));
-
-    return lb->second.endpoints;
+adhoc_t::resolve(const std::string& name) const -> service_description_t {
+    return m_remotes.apply([&](const remote_map_t& remotes) {
+        auto by_service_it = remotes.find(name);
+        if(by_service_it == remotes.end() || by_service_it->second.empty()) {
+            throw std::system_error(error::service_not_available);
+        }
+
+        // try to find local service
+        auto& services_by_uuid = by_service_it->second;
+        auto it = services_by_uuid.find({});
+        if(it != services_by_uuid.end()) {
+            COCAINE_LOG_DEBUG(m_log, "providing service using local actor");
+            return service_description_t{it->second.endpoints, it->second.protocol, it->second.version};
+        }
+
+        // roll the dice and choose random one
+        it = services_by_uuid.begin();
+        std::uniform_int_distribution<int> distribution(0, services_by_uuid.size() - 1);
+        std::advance(it, distribution(m_random_generator));
+
+        COCAINE_LOG_DEBUG(m_log, "providing service using remote actor", blackhole::attribute_list({
+            {"uuid", it->second.uuid}
+        }));
+
+        return service_description_t{it->second.endpoints, it->second.protocol, it->second.version};
+    });
 }
 
-size_t
-adhoc_t::consume(const std::string& uuid,
-                 const partition_t& name, const std::vector<asio::ip::tcp::endpoint>& endpoints)
-{
-    auto ptr = m_remotes.synchronize();
+auto
+adhoc_t::on_remote_service_exposed(const std::string& uuid,
+                                   const std::string& name,
+                                   unsigned int version,
+                                   const std::vector<asio::ip::tcp::endpoint>& endpoints,
+                                   const io::graph_root_t& protocol) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        COCAINE_LOG_DEBUG(m_log, "registering {}/{} destination with {:d} endpoints from {}",
+                          name, version, endpoints.size(), uuid);
+
+        auto& services_by_uuid = remotes[name];
+        bool inserted;
+        std::tie(std::ignore, inserted) = services_by_uuid.insert({uuid, remote_t{uuid, version, endpoints, protocol}});
+
+        if(!inserted) {
+            auto str = format("failed to add remote service {} located on {} to gateway: service already registered", name, uuid);
+            COCAINE_LOG_ERROR(m_log, str);
+            throw error_t(str);
```

Comment:
I still doubt that we need error_code for this specific case (I don't see any usage for it, and it IS invalid argument passed to function), but ok.